### PR TITLE
Improve culling and simulator responsiveness

### PR DESCRIPTION
**Pull request by rasmusgo**
* Handle all pending input events in `update_camera`- This fixes laggy input in the simulator when using a mouse with high refresh rate.
* Let the second camera follow the first. - This makes the interpolated pose culling work for the simulator.
* Replace shared camera frustum with separate frustums for the two cameras. - This removes the interpolated pose altogether.
* Move some redundant computations in `culling.comp` to the CPU so that the GPU code can be simplified.

### CODE REVIEW

**Comment by rasmusgo** [comment]
I think this should resolve https://github.com/leetvr/hotham/issues/275.

File: hotham-simulator/src/state.rs [hunk 1]
```diff
@@ -218,7 +218,7 @@ impl State {
         let movement_speed = 2f32 * dt;
         let mouse_sensitivity = 4f32 * dt;
 
-        if let Ok(input_event) = self.event_rx.as_ref()?.try_recv() {
+        while let Ok(input_event) = self.event_rx.as_ref()?.try_recv() {
```
Comment (kanerogers) [inline hotham-simulator/src/state.rs:221]: Good call

File: hotham-simulator/src/state.rs [hunk 2]
```diff
@@ -297,6 +297,7 @@ impl State {
         // self.spaces.get_mut(&right_hand).unwrap().position.z += z_delta;
         // self.spaces.get_mut(&right_hand).unwrap().position.x += x_delta;
 
+        self.view_poses[1] = self.view_poses[0];
```
Comment (kanerogers) [inline hotham-simulator/src/state.rs:300]: Nice

File: hotham/src/resources/render_context.rs [hunk 3]
```diff
@@ -229,26 +229,46 @@ impl RenderContext {
...
+        // We are only using four planes per camera. The near and far planes are not used.
+        // This link points to a paper describing the math behind these expressions:
+        // https://www.gamedevs.org/uploads/fast-extraction-viewing-frustum-planes-from-world-view-projection-matrix.pdf
+        let normalize_plane = |p: RowVector4<_>| p / p.columns(0, 3).norm();
         let cull_data = CullData {
-            frustum: frustum_projection * frustum_view,
+            clip_planes1: Matrix4::<_>::from_rows(&[
+                normalize_plane(left_frustum_from_world.row(3) + left_frustum_from_world.row(0)),
+                normalize_plane(left_frustum_from_world.row(3) - left_frustum_from_world.row(0)),
+                // normalize_plane(left_frustum_from_world.row(3) + left_frustum_from_world.row(1)),
```
Comment (kanerogers) [inline hotham/src/resources/render_context.rs:259]: Could you remove the commented out code?

File: hotham/src/resources/render_context.rs [hunk 4]
```diff
@@ -229,26 +229,46 @@ impl RenderContext {
...
+        let right_frustum_from_world = right_frustum_from_right_camera * right_camera_from_world;
 
+        // Normals of the clipping planes are pointing towards the inside of the frustum.
+        // We are only using four planes per camera. The near and far planes are not used.
+        // This link points to a paper describing the math behind these expressions:
+        // https://www.gamedevs.org/uploads/fast-extraction-viewing-frustum-planes-from-world-view-projection-matrix.pdf
+        let normalize_plane = |p: RowVector4<_>| p / p.columns(0, 3).norm();
         let cull_data = CullData {
-            frustum: frustum_projection * frustum_view,
+            clip_planes1: Matrix4::<_>::from_rows(&[
```
Comment (kanerogers) [inline hotham/src/resources/render_context.rs:256]: I'd prefer these be called something like `left_clip_plane` and `right_clip_plane`, just to be consistent with the other variables, both here and in the compute shader

File: hotham/src/resources/render_context.rs [hunk 5]
```diff
@@ -729,8 +749,9 @@ fn create_pipeline_layout(
 #[repr(C)]
 #[derive(Debug, Clone)]
 pub struct CullData {
-    /// View-Projection matrices (one per eye)
-    pub frustum: Matrix4<f32>,
+    /// Four clip planes per camera, one plane per row.
+    pub clip_planes1: Matrix4<f32>,
```
Comment (kanerogers) [inline hotham/src/resources/render_context.rs:753]: see above re: variable names

File: hotham/src/shaders/culling.comp [hunk 6]
```diff
@@ -27,48 +27,24 @@ layout(std430, set = 0, binding = 1) writeonly buffer DrawCommandsBuffer {
...
-    bool isVisible = check_is_visible(cullData.frustum, center, radius);
+    vec4 center4 = vec4(d.boundingSphere.xyz, 1);
+    vec4 negRadius4 = -d.boundingSphere.wwww;
+
+    // Treat as visible if at least one eye may see it
+    bool isVisible = any(bvec2(
+        all(greaterThan(cullData.clipPlanes1 * center4, negRadius4)),
+        all(greaterThan(cullData.clipPlanes2 * center4, negRadius4))
+    ));
     drawCommandsBuffer.drawCommands[id].instanceCount = isVisible ? 1 : 0;
```
Comment (kanerogers) [inline hotham/src/shaders/culling.comp:49]: Minor optimisation: cast the bool to a uint with uint(isVisible) - saves a couple instructions

**Review by kanerogers — CHANGES_REQUESTED** [review]
Overall this looks good! Just some minor changes suggested. :)

In future it would be nice to keep PRs like these separate (eg. the simulator responsiveness) but I understand how annoying that can be for small fixes

File: hotham/src/resources/render_context.rs (hunk 3 above)
Comment (rasmusgo) [reply hotham/src/resources/render_context.rs:259]: I left them in as a trace of what was omitted but it's a bit redundant given the comment above.

**Review by kanerogers — APPROVED** [review]
LGTM! Fantastic work as always @rasmusgo

File: hotham/src/shaders/culling.comp (hunk 6 above)
Comment (Ralith) [reply hotham/src/shaders/culling.comp]: FWIW, there's approximately zero chance the driver won't produce equivalent code for those.

Comment (kanerogers) [reply hotham/src/shaders/culling.comp]: On desktop land these assumptions are reasonable, but Qualcomm's documentation constantly warns about not assuming the driver's compiler will do what you want. 😛

**Pull request merged**